tests/gen/gen_predefined.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gen
+
+import (
+	"strings"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+)
+
+// GeneratePredefinedFromSDL generates documents for GraphQL SDL from a predefined list
+// of docs that might include nested docs.
+// The SDL is parsed to get the list of fields, and the docs
+// are created with the fields parsed from the SDL.
+// This allows us to have only one large list of docs with predefined
+// fields, and create SDLs with different fields from it.
+func GeneratePredefinedFromSDL(gqlSDL string, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+// GeneratePredefined generates documents from a predefined list
+// of docs that might include nested docs.
+func GeneratePredefined(defs []client.CollectionDefinition, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, col := range defs {
+		typeDefs[col.Description.Name] = col
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+type docGenerator struct {
+	types map[string]client.CollectionDefinition
+}
+
+// toRequestedDoc removes the fields that are not in the SDL of the collection.
```

Comment:
todo: This comment does not make sense to me, do you mean `the schema of the collection. `/`the collection`s schema.`?  And if so, how did these fields find their way on to the document in the first place, if they are not part of that collection/schema why were they added?

Reply:
There is an example of it in README.md. You can provide a large predefined set to `gen.GeneratePredefined` with document that might contain properties that are not defined in the collection. They will be filtered out.

Reply:
Okay fair enough, that is however unexpected - could you please amend this code comment so that it contains this information?  Or change the code so that fields that do not exist are not added to the doc in the first place. 

And please change the wording to something similar to the original suggestion - collections do not have SDLs, they have a schema.

Reply:
> Or change the code so that fields that do not exist are not added to the doc in the first place.

Code doesn't add the fields. They are provided by the user, or a test. You can take a look at how it's used here tests/integration/index/query_with_relation_filter_test.go

Reply:
Ah okay - cheers, I think I get it now, this code is really removing the fields from the input 'seed' documents, not the resultant output?

If that's correct it might be worth noting that in the function documentation, something like:

```
// toRequestedDoc removes the fields that are not in the schema of the collection.
//
// This is typically called on user/test provided seed documents to remove any non-existent fields before generating
// documents from them. [And then perhaps clarify whether the input set is mutated, as I'm not sure]
```

Reply:
done